### ISSUE TITLE
Overly aggressive inlining of Date expressions

### ISSUE THREAD

**Issue by mfikes** [google/closure-compiler#3079]
Consider this JavaScript code:

```javascript
console.log(
  (function() {
    var start = new Date().getTime();
    var end = new Date().getTime();
    return end - start;
  })()
);
```

Passing this through advanced optimizations yields

```javascript
console.log((new Date).getTime()-(new Date).getTime());
```

which has different semantics: The unoptimized version will calculate a an end date after the start date, and the difference will be non-negative. The optimized version will evaluate left to right, calculating the left date first, followed by the right date, producing non-positive values.

Example, using Nashorn:

```
jjs> (new Date).getTime()-(new Date).getTime()
-3
```

**Comment by mfikes** [comment]
@lazarljubenovic Yep, it is the result of a different optimization. This was found via ClojureScript, which can dynamically generate code (via a macro) to measure the length of time needed to evaluate an expression. If the thing you are measuring turns out to be trivial and gets (correctly) elided during advanced, then the inlining issue crops up.

**Comment by concavelenz** [comment]
This is an interesting question.  We generally code motion of side-effect free expressions.  A side effect in either the "start" or "end" expressions would prevent the result we see here.  However as neither expression is considered to have side-effects (which is true), they get inlined.  However the reality is the side-effect is occurring somewhere the compiler can't see it (time is passing and the clock is running).   A simple fix would be to remove `@nosideeffects` from the Date constructor and the Date.now static method but that would also make those calls unremovable.

**Comment by vemv** [comment]
Perhaps it would help to have a distinction between side-effect-free and referentially-pure functions.

That way, best-case scenario, some optimizations could be preserved, while the incorrect ones would be discarded.

**Comment by EatingW** [comment]
Created internal issue http://b/133856418
